File: frontend_dashboard/tests/run_tests.js
import { spawn } from 'child_process';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.resolve(__dirname, '..');

// Helper to format date
const formatDate = (date) => {
  const pad = (n) => n.toString().padStart(2, '0');
  const yyyy = date.getFullYear();
  const mm = pad(date.getMonth() + 1);
  const dd = pad(date.getDate());
  const hh = pad(date.getHours());
  const min = pad(date.getMinutes());
  const ss = pad(date.getSeconds());
  return `${yyyy}${mm}${dd}_${hh}${mm}${ss}`;
};

const timestamp = formatDate(new Date());
const logsDir = path.join(__dirname, 'logs');
const reportsDir = path.join(__dirname, 'reports');
const logFile = path.join(logsDir, `test_run_${timestamp}.log`);
const reportFile = path.join(reportsDir, `test_report_${timestamp}.md`);
const tempJsonFile = path.join(__dirname, `temp_results_${timestamp}.json`);

// Ensure directories exist
if (!fs.existsSync(logsDir)) fs.mkdirSync(logsDir, { recursive: true });
if (!fs.existsSync(reportsDir)) fs.mkdirSync(reportsDir, { recursive: true });

console.log(`Starting test run...`);
console.log(`Log file: ${logFile}`);
console.log(`Report file: ${reportFile}`);

// Create write stream for log file
const logStream = fs.createWriteStream(logFile, { flags: 'a' });

// Run Vitest
// We use two reporters: 'verbose' for console/log output, and 'json' for generating the report
// Note: We need to be careful about how we capture output.
const vitest = spawn(/^win/.test(process.platform) ? 'npx.cmd' : 'npx', [
  'vitest',
  'run',
  '--reporter=verbose',
  '--reporter=json',
  `--outputFile.json=${tempJsonFile}`
], {
  cwd: projectRoot,
  env: { ...process.env, CI: 'true' }, // Force CI mode to avoid watch mode and colors if needed
  stdio: ['inherit', 'pipe', 'pipe']
});

// Pipe output to console and log file
vitest.stdout.on('data', (data) => {
  process.stdout.write(data);
  logStream.write(data);
});

vitest.stderr.on('data', (data) => {
  process.stderr.write(data);
  logStream.write(data);
});

vitest.on('close', (code) => {
  logStream.end();
  console.log(`\nTest run finished with exit code ${code}`);
  
  generateReport(code);
});

function generateReport(exitCode) {
  if (!fs.existsSync(tempJsonFile)) {
    console.error('Error: JSON result file not found. Cannot generate report.');
    return;
  }

  try {
    const rawData = fs.readFileSync(tempJsonFile, 'utf-8');
    const results = JSON.parse(rawData);
    
    // Calculate stats
    const totalTests = results.numTotalTestSuites; // This is suites, let's count tests
    let totalTestCases = 0;
    let passedTests = 0;
    let failedTests = 0;
    let skippedTests = 0;
    let totalDuration = 0;

    // Traverse results
    const processSuite = (suite) => {
        if (suite.assertionResults) {
            suite.assertionResults.forEach(test => {
                totalTestCases++;
                if (test.status === 'passed') passedTests++;
                else if (test.status === 'failed') failedTests++;
                else if (test.status === 'skipped' || test.status === 'pending') skippedTests++;
                
                // Duration is not always directly on test object in some versions, check
                // Usually it has 'duration' in ms
            });
        }
    };
    
    // Recursive traversal if needed, but vitest json structure is usually flat on testResults
    if (results.testResults) {
        results.testResults.forEach(suite => {
            // totalDuration += suite.endTime - suite.startTime; // Approximate
            if (suite.assertionResults) {
                 suite.assertionResults.forEach(test => {
                    totalTestCases++;
                    if (test.status === 'passed') passedTests++;
                    else if (test.status === 'failed') failedTests++;
                    else if (test.status === 'skipped' || test.status === 'pending') skippedTests++;
                 });
            }
        });
    }
    
    // Use top level stats if available
    totalTestCases = results.numTotalTests || totalTestCases;
    passedTests = results.numPassedTests || passedTests;
    failedTests = results.numFailedTests || failedTests;
    // skipped is not always top level
    
    // If stats are missing, we might need to count them manually from testResults
    // Let's rely on what we can find.
    
    const statusIcon = exitCode === 0 ? '✅ PASSED' : '❌ FAILED';
    const passRate = totalTestCases > 0 ? ((passedTests / totalTestCases) * 100).toFixed(2) + '%' : 'N/A';
    
    const startTime = new Date(results.startTime);
    // Vitest JSON might not have total duration at top level, use startTime
    const duration = (Date.now() - startTime.getTime()) / 1000;

    let markdown = `# Test Execution Report

**Run ID**: ${timestamp}
**Status**: ${statusIcon}
**Date**: ${new Date().toLocaleString()}

## 1. Overview
| Metric | Value |
|--------|-------|
| Total Tests | ${totalTestCases} |
| Passed | ${passedTests} |
| Failed | ${failedTests} |
| Skipped | ${totalTestCases - passedTests - failedTests} |
| Pass Rate | ${passRate} |
| Duration | ${duration.toFixed(2)}s |

## 2. Detailed Results

`;

    if (results.testResults) {
        results.testResults.forEach(suite => {
            const suiteName = path.relative(projectRoot, suite.name);
            markdown += `### 📄 ${suiteName}\n\n`;
            markdown += `| Status | Test Case | Duration | Error |\n`;
            markdown += `| :---: | :--- | :---: | :--- |\n`;
            
            if (suite.assertionResults) {
                suite.assertionResults.forEach(test => {
                    const icon = test.status === 'passed' ? '✅' : (test.status === 'failed' ? '❌' : '⚠️');
                    const duration = test.duration ? `${test.duration}ms` : '-';
                    let errorMsg = '';
                    if (test.status === 'failed' && test.failureMessages) {
                         errorMsg = test.failureMessages.join(' ').replace(/\n/g, '<br>').replace(/\|/g, '\\|');
                    }
                    markdown += `| ${icon} | ${test.title} | ${duration} | ${errorMsg} |\n`;
                });
            }
            markdown += '\n';
        });
    }

    fs.writeFileSync(reportFile, markdown);
    console.log(`Report generated: ${reportFile}`);

    // Cleanup
    fs.unlinkSync(tempJsonFile);

  } catch (err) {
    console.error('Error generating report:', err);
  }
}

File: backend_system/tests/run_tests.py
import sys
import os
import subprocess
import xml.etree.ElementTree as ET
from datetime import datetime

def run_tests():
    # Setup directories
    script_dir = os.path.dirname(os.path.abspath(__file__))
    
    # Check if we are inside the 'tests' directory
    if os.path.basename(script_dir) == 'tests':
        project_root = os.path.dirname(script_dir)
        tests_dir = script_dir
    else:
        project_root = script_dir
        tests_dir = os.path.join(project_root, "tests")
        
    logs_dir = os.path.join(tests_dir, "logs")
    reports_dir = os.path.join(tests_dir, "reports")
    
    os.makedirs(logs_dir, exist_ok=True)
    os.makedirs(reports_dir, exist_ok=True)
    
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    log_file = os.path.join(logs_dir, f"test_run_{timestamp}.log")
    report_file = os.path.join(reports_dir, f"test_report_{timestamp}.md")
    xml_report_file = os.path.join(reports_dir, f"temp_result_{timestamp}.xml")
    
    print(f"Starting test run at {timestamp}...")
    print(f"Project root: {project_root}")
    print(f"Logs will be saved to: {log_file}")
    print(f"Report will be saved to: {report_file}")
    
    # Command to run pytest
    # -v: verbose output
    # --junitxml: generate XML report for parsing
    
    # Allow passing specific test paths or arguments from command line
    test_args = sys.argv[1:] if len(sys.argv) > 1 else ["tests/"]
    
    cmd = [
        sys.executable, "-m", "pytest", 
        *test_args,
        f"--junitxml={xml_report_file}",
        "-v"
    ]
    
    exit_code = 0
    
    try:
        with open(log_file, "w", encoding="utf-8") as f:
            # Write header to log
            f.write(f"Test Run: {timestamp}\n")
            f.write(f"Command: {' '.join(cmd)}\n")
            f.write("-" * 80 + "\n")
            
            # Run pytest, capturing output
            process = subprocess.Popen(
                cmd, 
                stdout=subprocess.PIPE, 
                stderr=subprocess.STDOUT,
                text=True,
                cwd=project_root,
                env=os.environ.copy()
            )
            
            # Stream output to console and file
            for line in process.stdout:
                sys.stdout.write(line)
                f.write(line)
                
            exit_code = process.wait()
            
        # Generate Markdown Report from XML
        if os.path.exists(xml_report_file):
            generate_markdown_report(xml_report_file, report_file, timestamp, exit_code)
            # Cleanup XML file after report generation
            os.remove(xml_report_file)
        else:
            print("Error: JUnit XML report was not generated.")
            
    except Exception as e:
        print(f"An error occurred during test execution: {e}")
        exit_code = 1
        
    return exit_code

def generate_markdown_report(xml_file, md_file, timestamp, exit_code):
    try:
        tree = ET.parse(xml_file)
        root = tree.getroot()
        
        # Test Suite Summary
        # pytest junitxml usually puts a root <testsuites> or single <testsuite>
        total_tests = 0
        failures = 0
        errors = 0
        skipped = 0
        time_taken = 0.0
        
        test_cases = []
        
        # Helper to process a testsuite element
        def process_suite(suite):
            nonlocal total_tests, failures, errors, skipped, time_taken
            total_tests += int(suite.get('tests', 0))
            failures += int(suite.get('failures', 0))
            errors += int(suite.get('errors', 0))
            skipped += int(suite.get('skipped', 0))
            time_taken += float(suite.get('time', 0))
            
            for tc in suite.iter('testcase'):
                status = "PASS"
                details = ""
                # Check for failure, error, skipped elements
                fail_node = tc.find('failure')
                error_node = tc.find('error')
                skip_node = tc.find('skipped')
                
                if fail_node is not None:
                    status = "FAIL"
                    details = fail_node.get('message', '') or fail_node.text or "Failure"
                elif error_node is not None:
                    status = "ERROR"
                    details = error_node.get('message', '') or error_node.text or "Error"
                elif skip_node is not None:
                    status = "SKIP"
                    details = skip_node.get('message', '') or skip_node.text or "Skipped"
                
                test_cases.append({
                    "name": tc.get('name'),
                    "classname": tc.get('classname'),
                    "time": float(tc.get('time', 0)),
                    "status": status,
                    "details": details
                })

        if root.tag == 'testsuites':
            for suite in root.iter('testsuite'):
                process_suite(suite)
        else:
            process_suite(root)
        
        passed = total_tests - failures - errors - skipped
        pass_rate = (passed / total_tests * 100) if total_tests > 0 else 0
        
        status_emoji = "✅ PASSED" if exit_code == 0 else "❌ FAILED"
        
        with open(md_file, "w", encoding="utf-8") as f:
            f.write(f"# Test Execution Report\n\n")
            f.write(f"**Run ID**: {timestamp}\n")
            f.write(f"**Date**: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
            f.write(f"**Status**: {status_emoji}\n\n")
            
            f.write("## 1. Overview\n\n")
            f.write(f"| Metric | Value |\n")
            f.write(f"| :--- | :--- |\n")
            f.write(f"| **Total Tests** | {total_tests} |\n")
            f.write(f"| **Passed** | {passed} |\n")
            f.write(f"| **Failed** | {failures} |\n")
            f.write(f"| **Errors** | {errors} |\n")
            f.write(f"| **Skipped** | {skipped} |\n")
            f.write(f"| **Pass Rate** | {pass_rate:.2f}% |\n")
            f.write(f"| **Total Time** | {time_taken:.2f}s |\n\n")
            
            f.write("## 2. Detailed Results\n\n")
            f.write("| Test Case | Module | Time (s) | Status | Details |\n")
            f.write("| :--- | :--- | :--- | :--- | :--- |\n")
            
            for tc in test_cases:
                status_icon = "✅" if tc['status'] == "PASS" else "❌" if tc['status'] in ["FAIL", "ERROR"] else "⚠️"
                # Format details to be table-friendly (remove newlines, limit length)
                raw_details = str(tc['details']).strip()
                details_preview = (raw_details[:100] + '...') if len(raw_details) > 100 else raw_details
                details_preview = details_preview.replace("\n", " ").replace("|", "\|")
                
                f.write(f"| {tc['name']} | {tc['classname']} | {tc['time']:.3f} | {status_icon} {tc['status']} | {details_preview} |\n")
                
        print(f"\nReport generated successfully: {md_file}")
        
    except Exception as e:
        print(f"\nFailed to generate report: {e}")

if __name__ == "__main__":
    sys.exit(run_tests())
